### PR TITLE
Feature/boto client config

### PR DESCRIPTION
**Pull request by jona62**
*Issue #, if available:* N/A

*Description of changes:*

This PR enhances the MemorySessionManager with improved boto3 client configuration support and timestamp precision handling:

**Commit 1: Add support for boto_client_config**
- Added `boto_client_config` parameter to MemorySessionManager constructor to allow custom botocore configuration
- Implemented intelligent config merging that preserves user-provided settings while adding SDK user agent
- Enhanced user agent handling to append "bedrock-agentcore-sdk" to existing user agents or set as default
- Added comprehensive test coverage (331 new lines) including tests for:
  - Custom retry configurations
  - Connection pool settings
  - User agent preservation and merging
  - Integration with boto3 sessions
  - Edge cases and parameter validation

**Commit 2: Override timestamp serializer for microsecond granularity**
- Implemented custom timestamp serialization to preserve microsecond precision when sending datetime objects to AgentCore Memory service
- Refactored MemorySessionManager initialization into modular helper methods:
  - `_validate_and_resolve_region()` for region consistency validation
  - `_build_client_config()` for boto3 client configuration building
  - `_configure_timestamp_serialization()` for timestamp precision setup
- Fixed region validation logic to properly handle session region types
- Improved code organization and maintainability through method extraction

These changes provide:
1. **Enhanced Configuration Flexibility**: Users can now customize boto3 client behavior (retries, timeouts, connection pools) while maintaining SDK functionality
2. **Improved Timestamp Precision**: Datetime objects now preserve microsecond precision in API calls, critical for time-sensitive memory operations
3. **Better Code Structure**: Modular initialization methods improve readability and testability
4. **Comprehensive Testing**: Extensive test coverage ensures reliability across various configuration scenarios

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/bedrock_agentcore/memory/session.py [hunk 1]
```diff
@@ -132,6 +138,71 @@ def __init__(
...
+        Returns:
+            The resolved region name to use
+
+        Raises:
+            ValueError: If region_name conflicts with session region
+        """
+        session_region = session.region_name
+
+        # Validate region consistency if both are provided
+        if region_name and session_region and isinstance(session_region, str) and region_name != session_region:
```
Comment (bergjaak) [inline src/bedrock_agentcore/memory/session.py:157]: I think it's more standard to have a priority order and take whichever is highest priority. I think AWS_REGION env var usually is first priority

File: src/bedrock_agentcore/memory/session.py [hunk 2]
```diff
@@ -132,6 +138,71 @@ def __init__(
...
+
+    def _configure_timestamp_serialization(self) -> None:
+        """Configure the boto3 client to serialize timestamps with microsecond precision.
+
+        This method overrides the default timestamp serialization to preserve microsecond
+        precision when sending datetime objects to the AgentCore Memory service.
+        """
+        original_serialize_timestamp = self._data_plane_client._serializer._serializer._serialize_type_timestamp
+
+        def serialize_timestamp_with_microseconds(serialized, value, shape, name):
```
Comment (bergjaak) [inline src/bedrock_agentcore/memory/session.py]: it's actually milliseconds, right?

**Review by bergjaak — DISMISSED** [review]

File: src/bedrock_agentcore/memory/session.py (hunk 1 above)
Comment (jona62) [reply src/bedrock_agentcore/memory/session.py:157]: If you don't specify a `region_name` param, boto3 checks for the `AWS_REGION` and `AWS_DEFAULT_REGION` environment variables. If either of these are set, their value will be used as the default region. Then lastly it checks the AWS configuration file `~/.aws/config`

File: src/bedrock_agentcore/memory/session.py [hunk 3]
```diff
@@ -132,6 +138,70 @@ def __init__(
...
+                new_user_agent = sdk_user_agent
+            return boto_client_config.merge(BotocoreConfig(user_agent_extra=new_user_agent))
+        else:
+            return BotocoreConfig(user_agent_extra=sdk_user_agent)
+
+    def _configure_timestamp_serialization(self) -> None:
+        """Configure the boto3 client to serialize timestamps as float values.
+
+        This method overrides the default timestamp serialization to convert datetime objects
+        to float timestamps (seconds since Unix epoch) which preserves microsecond precision
```
Comment (bergjaak) [inline src/bedrock_agentcore/memory/session.py]: millisecond?

**Review by bergjaak — APPROVED** [review]

File: src/bedrock_agentcore/memory/session.py (hunk 2 above)
Comment (jona62) [reply src/bedrock_agentcore/memory/session.py]: You're correct, renamed the function and updated the PR
